### ISSUE TITLE
Unix Timestamp as DateTime Format Specifier

### ISSUE THREAD

**Issue by hannasm** [dotnet/runtime#43562]
How about we add something to handle Unix Timestamps in the DateTime _standard_ or _custom_ format strings. 

It would be great for me if i could handle a prefix and suffix like i get on images downloaded from text messages on my android cell phone like 'Message_1602994376381.jpg'

```c#
[DateTime]::ParseExact('Message_1602994376381.jpg', '\M\e\s\s\a\g\e\_X\.\j\p\g', $filename, [System.Globalization.CultureInfo]::InvariantCulture)
```

In this code snippet i am implying that the 'X' be handled as a unix timestamp.

I am not aware of a way for the date time parser to handle prefix / suffix on a standard format specifier,and that makes my use case problematic with 'X' as a standard format specifier. Custom format specifiers can handle prefixes / suffixes but don't handle variadic lengths, and the unix format specifier can vary between  -62135596800000 -and 253402300799999 which implies anywhere between 0 and 15 digits. 

Something as described above would be ideal in my use case. However, adding an 'X' standard format specifier would probably be the easiest addition to the existing framework. Adding 'X' as a custom format specifier requiring specific number of digits to be specified, might be annoying to use but, could help get the job done.

Considering i need this right now i will have to implement a workaround myself. This would be a fairly boring improvement to the API surface. I would even be happy to submit a patch if an acceptable approach could be agreed upon. 

Add something to the documentation here:

https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings
https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings

That ends up calling the function here:

https://docs.microsoft.com/en-us/dotnet/api/system.datetimeoffset.fromunixtimemilliseconds

**Comment by Clockwork-Muse** [comment]
> I am not aware of a way for the date time parser to handle prefix / suffix on a standard format specifier,and that makes my use case problematic with 'X' as a standard format specifier. Custom format specifiers can handle prefixes / suffixes but don't handle variadic lengths, and the unix format specifier can vary between -62135596800000 -and 253402300799999 which implies anywhere between 0 and 15 digits.

Yes, because for _formatting_ the intent is that any extra characters come from the external formatting string, and you just use the formatting specifier.  Parsing has much the same intent - the extra characters are meant to handle some oddball formats, not extraction as you're doing here.  So standard formats are likely to always reject extra characters.

Note that custom and standard format/parse strings _can_ (and often do) have varying lengths - consider month/day-of-week names (and all the various localizations), plus some formats don't prepend zeroes.

> It would be great for me if i could handle a prefix and suffix like i get on images downloaded from text messages on my android cell phone like 'Message_1602994376381.jpg'

Arguably, this is a misuse of the parsing code, and you should be parsing this with `Regex` or a cut down parser (there are various libraries that would help craft this stuff, but a compiled regex for a simple case is likely to work well).

For your particular case, I'd also check to see if any of the EXIF or file data has been set, you may not have to parse anything.

**Comment by hannasm** [comment]
I wouldn't disagree with anything about that, but I also still believe a
Unix timestamp qualifies as a DateTime format, and should be supported as
such in the format specifiers and parser.

Given an extraction use case is out of scope a standard format specifier is
the only sensible choice.

Something like DateTime.ToString('X') in the format provider would convert
to a Unix timestamp and a ParseExact('X') would read one.

**Comment by benaadams** [comment]
Unix timestamp format is a simple numeric `long`?

There are methods on `DateTimeOffset` for converting to and from that `long`
```csharp
struct DateTimeOffset
{
    DateTimeOffset FromUnixTimeMilliseconds(long milliseconds);
    DateTimeOffset FromUnixTimeSeconds(long seconds);
    long ToUnixTimeSeconds();
    long ToUnixTimeMilliseconds();
}
```

**Comment by IanKemp** [comment]
> I wouldn't disagree with anything about that, but I also still believe a Unix timestamp qualifies as a DateTime format, and should be supported as such in the format specifiers and parser. Given an extraction use case is out of scope a standard format specifier is the only sensible choice. Something like DateTime.ToString('X') in the format provider would convert to a Unix timestamp and a ParseExact('X') would read one.

This is long overdue to be added. Will see if I can throw up a PR for it.
